File: Java/com/example/modules/GaussSolver.java
package com.example.modules;

import java.util.Arrays;

public class GaussSolver {

	public static double[] solve(double[][] matrix, double[] solutions) {
		final int n = matrix.length;
		if (n == 0 || matrix[0].length != n || solutions.length != n) {
			throw new IllegalArgumentException(
				"Invalid matrix or vector dimensions."
			);
		}

		// Создаем расширенную матрицу
		double[][] augmentedMatrix = new double[n][n + 1];
		for (int i = 0; i < n; i++) {
			System.arraycopy(matrix[i], 0, augmentedMatrix[i], 0, n);
			augmentedMatrix[i][n] = solutions[i];
		}

		// Прямой ход (Приведение к верхне-треугольному виду)
		for (int k = 0; k < n; k++) {
			// Поиск ведущего элемента (для стабильности)
			int pivotRow = k;
			for (int i = k + 1; i < n; i++) {
				if (
					Math.abs(augmentedMatrix[i][k]) >
					Math.abs(augmentedMatrix[pivotRow][k])
				) {
					pivotRow = i;
				}
			}

			// Обмен строк
			double[] temp = augmentedMatrix[k];
			augmentedMatrix[k] = augmentedMatrix[pivotRow];
			augmentedMatrix[pivotRow] = temp;

			if (Math.abs(augmentedMatrix[k][k]) < 1e-10) {
				throw new ArithmeticException(
					"Matrix is singular or nearly singular."
				);
			}

			for (int i = k + 1; i < n; i++) {
				double factor = augmentedMatrix[i][k] / augmentedMatrix[k][k];
				for (int j = k; j < n + 1; j++) {
					augmentedMatrix[i][j] -= factor * augmentedMatrix[k][j];
				}
			}
		}

		// Обратный ход
		double[] x = new double[n];
		for (int i = n - 1; i >= 0; i--) {
			double sum = 0.0;
			for (int j = i + 1; j < n; j++) {
				sum += augmentedMatrix[i][j] * x[j];
			}
			x[i] = (augmentedMatrix[i][n] - sum) / augmentedMatrix[i][i];
		}
		return x;
	}

	public static void run() {
		double[][] matrix = {
			{ 0.91, -0.04, 0.21, -18.0 },
			{ 0.25, -1.23, -0.23, -0.09 },
			{ -0.21, -0.23, 0.80, -0.13 },
			{ 0.15, -1.31, 0.06, -1.04 },
		};
		double[] solutions = { -1.24, -1.04, 2.56, 0.91 };

		try {
			double[] result = solve(matrix, solutions);
			System.out.println("Решение найдено методом Гаусса:");
			for (int i = 0; i < result.length; i++) {
				System.out.printf("x%d = %.5f%n", i + 1, result[i]);
			}
		} catch (Exception e) {
			System.err.println("Ошибка: " + e.getMessage());
		}
	}
}

File: C++/src/modules/gauss_solver.cpp
#include "../../include/modules/gauss_solver.hpp"

#include <algorithm>
#include <cmath>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <vector>

std::vector<double> solve(std::vector<std::vector<double>> matrix,
						  const std::vector<double> &solutions) {
	const size_t n = matrix.size();
	if (n == 0 || matrix.at(0).size() != n || solutions.size() != n) {
		throw std::invalid_argument("Invalid matrix or vector dimensions.");
	}

	for (size_t i = 0; i < n; ++i) {
		matrix.at(i).push_back(solutions.at(i));
	}

	for (size_t k = 0; k < n; ++k) {
		size_t pivot_row = k;
		for (size_t i = k + 1; i < n; ++i) {
			if (std::abs(matrix.at(i).at(k)) > std::abs(matrix.at(pivot_row).at(k))) {
				pivot_row = i;
			}
		}

		if (pivot_row != k) {
			std::swap(matrix.at(k), matrix.at(pivot_row));
		}

		if (std::abs(matrix.at(k).at(k)) < 1e-10) {
			throw std::runtime_error("Matrix is singular or nearly singular.");
		}

		for (size_t i = k + 1; i < n; ++i) {
			const double factor = matrix.at(i).at(k) / matrix.at(k).at(k);
			for (size_t j = k; j < n + 1; ++j) {
				matrix.at(i).at(j) -= factor * matrix.at(k).at(j);
			}
		}
	}

	std::vector<double> x(n);
	for (int i = n - 1; i >= 0; --i) {
		double sum = 0.0;
		for (size_t j = i + 1; j < n; ++j) {
			sum += matrix.at(i).at(j) * x.at(j);
		}
		x.at(i) = (matrix.at(i).at(n) - sum) / matrix.at(i).at(i);
	}

	return x;
}

void runGaussSolver() {
	std::vector<std::vector<double>> matrix = {{0.91, -0.04, 0.21, -18.0},
											   {0.25, -1.23, -0.23, -0.09},
											   {-0.21, -0.23, 0.80, -0.13},
											   {0.15, -1.31, 0.06, -1.04}};

	std::vector<double> solutions = {-1.24, -1.04, 2.56, 0.91};

	try {
		std::vector<double> x = solve(matrix, solutions);

		std::cout << "Решение найдено методом Гаусса:" << std::endl;
		std::cout << std::fixed << std::setprecision(5);
		for (size_t i = 0; i < x.size(); ++i) {
			std::cout << "x" << i + 1 << " = " << x.at(i) << std::endl;
		}
	} catch (const std::exception &e) {
		std::cerr << "Error: " << e.what() << std::endl;
		return;
	}
}
